jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java

```diff
@@ -2,34 +2,133 @@
+        }
+
+        public FileInfo(long lastChangeTs, Throwable exception) {
+            this(lastChangeTs, null, exception);
+        }
+
+        public FileInfo(long lastChangeTs, String fileName, Throwable exception) {
+            this.lastChangeTs = lastChangeTs;
+            this.fileName = fileName;
+            this.exception = exception;
+        }
+
+    }
+
+    public static void premain(String agentArgument, Instrumentation instrumentation) throws Exception {
+        final String[] args = agentArgument.split(":");
+        if (args.length < 2) {
+            System.err.println("Usage: -javaagent:/path/to/JavaAgent.jar=<port>:<yaml configuration file>[:<config watch timeout - default 30s>]");
+            System.exit(1);
+        }
+
+        long timeout = 30l;
+
+        if (args.length > 2) {
+            timeout = Long.parseLong(args[2]);
+        }
+
+        final CountDownLatch jmxDone = new CountDownLatch(1);
+        watchTimer.scheduleAtFixedRate(new JmxConfigurer(jmxDone, new File(args[1])), 0l, TimeUnit.SECONDS.toMillis(timeout));
+        DefaultExports.initialize();
+
+        try {
+            jmxDone.await(10, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Failed waiting for JMX registry", e);
+        }
+
+        FileInfo info = lastChange.get();
+        if (info.exception != null) {
+            throw new IllegalStateException("Failed to initialise agent", info.exception);
+        }
+
+        int port = Integer.parseInt(args[0]);
+        server = new Server(port);
+        QueuedThreadPool pool = new QueuedThreadPool();
+        pool.setDaemon(true);
+        server.setThreadPool(pool);
+        ServletContextHandler context = new ServletContextHandler();
+        context.setContextPath("/");
+        server.setHandler(context);
+        context.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
+        server.start();
+    }
+
+    static class JmxConfigurer extends TimerTask {
+        final CountDownLatch startLatch;
+        final File fileToWatch;
+        final ConcurrentCollectorWrapper wrapper = new ConcurrentCollectorWrapper();
+
+        JmxConfigurer(CountDownLatch startLatch, File fileToWatch) {
+            this.startLatch = startLatch;
+            this.fileToWatch = fileToWatch;
+        }
+
+        @Override
+        public void run() {
+            FileReader reader = null;
+            FileInfo info = null;
+            try {
+                FileInfo lastLoad = lastChange.get();
+                long lastModified = fileToWatch.lastModified();
+                if (lastLoad != null && lastLoad.lastChangeTs == lastModified) {
+                    return;
+                }
+
+                reader = new FileReader(fileToWatch);
+                JmxCollector collector = new JmxCollector(reader);
+
+                // we don't care about double registering, since wrapper returns itself on set
```

Comment:
Registering a collector twice is not something you should be doing, it should happen only once.